### PR TITLE
maxflow

### PR DESCRIPTION
**Pull request by tommytgraph**

### CODE REVIEW

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 1]
```diff
@@ -0,0 +1,186 @@
+CREATE QUERY tg_maxflow(VERTEX source, VERTEX sink, Set<STRING> v_type, SET<STRING> e_type, 
+       SET<STRING> reverse_e_type, STRING cap_attr, STRING cap_type, FLOAT min_flow_threshhold = 0.001, 
+       BOOL print_accum = TRUE, BOOL display_edges = TRUE, BOOL spit_to_file = FALSE, 
+       STRING file_path = "/home/tigergraph/tg_query_output.csv"){
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:4]: set default file_path as `STRING file_path = ""`
this aligns better with other algorithms in the library

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  MinAccum<FLOAT> @@minflow;
+  OrAccum @is_visited, @@is_found;
+  BOOL minimum_reached = FALSE;
+	FILE f(file_path);
+  @@cap_threshold = min_flow_threshhold;
+         
+  IF cap_type NOT IN ("UINT", "INT", "FLOAT", "DOUBLE") THEN
+		PRINT "wt_type must be UINT, INT, FLOAT, or DOUBLE" AS errMsg;
+		RETURN;
+	END;
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:35]: think this type checking plus the CASE statement is a good way to make `e.getAttr(ATTRIBUTE_NAME, TYPE_STRING)` take different attributes type input; given TYPE_STRING still only takes a static string in 3.2
should consider adopting a similar approach for future algorithms in the library @victor-gsl

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+					    fl = e.getAttr(cap_attr, "INT")
+            WHEN "FLOAT" THEN
+              fl = e.getAttr(cap_attr, "FLOAT")
+            WHEN "DOUBLE" THEN
+              fl = e.getAttr(cap_attr, "DOUBLE")
+          END,
+          @@flow_accum += (s, t -> 0),
+          IF s == source THEN @@cap_threshold += fl END;
+
+  @@cap_threshold = pow(3, float_to_int(log(@@cap_threshold)/log(3)));
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:56]: add some comments on how cap_threshold is calculated/ used?
a reference to the capacity scaling algorithm will be helpful
you can do it as general information comments at the beginning or when the cap_threshold is used

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                               @@flow_accum += (s.@traceback.top().prev, s -> @@minflow)
+                          ELSE @@flow_accum += (s, s.@traceback.top().prev -> -@@minflow)
+                         END;
+    END;
+  END;
+  ##### Output #####
+	IF file_path != "" THEN
+    f.println("Maxflow: " + to_string(@@maxflow));
+	  f.println("From","To","Flow");
+	END;
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:166]: please add an empty line before `##### Output #####` and fix the indentation

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 5]
```diff
@@ -0,0 +1,186 @@
+CREATE QUERY tg_maxflow(VERTEX source, VERTEX sink, Set<STRING> v_type, SET<STRING> e_type, 
+       SET<STRING> reverse_e_type, STRING cap_attr, STRING cap_type, FLOAT min_flow_threshhold = 0.001, 
+       BOOL print_accum = TRUE, BOOL display_edges = TRUE, BOOL spit_to_file = FALSE,
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:3]: parameter `spit_to_file` was never used in the query
please remove it if not needed

File: algorithms/Path/maxflow/tg_maxflow.gsql [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+       BOOL print_accum = TRUE, BOOL display_edges = TRUE, BOOL spit_to_file = FALSE, 
+       STRING file_path = "/home/tigergraph/tg_query_output.csv"){ 
+  
+  /*
+   Maxflow algorithm. Finds the maximum amount of flow that source can pushes to sink
+
+   source: start vertex to  every other vertex.     cap_type: capacity data type (UINT, INT,FLOAT,DOUBLE)
+   sink:   end vertex                               min_flow_threshhold: smallest epsilon for flow
+   v_type: vertex types to traverse                 print_accum: print JSON output
+   e_type: edge types to traverse                   display_edges: output edges for visualization
```
Comment (Boyu1997) [inline algorithms/Path/maxflow/tg_maxflow.gsql:12]: comment for `display_edges` should clarify this only works when `print_accum = TRUE`

**Review by Boyu1997 — CHANGES_REQUESTED** [review]
Hi Tommy,
Great job implementing the MaxFlow algorithms! Some minor changes before the merge.

**Review by wyatt-joyner-tg — APPROVED** [review]
I see that you got all the style changes in! Great work, Tommy.

**Comment by Boyu1997** [comment]
@tommytgraph @wyatt-joyner-tg :rocket::rocket::rocket:

**Pull request merged**